clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -538,63 +530,54 @@ public void onComplete(ClientResponse response) {
 
 
     /*
-     * A threadsafe helper class to hold requests per node that has not been sent yet
+     * A threadsafe helper class to hold requests per node that have not been sent yet
      */
     private final static class UnsentRequests {
-        private final Map<Node, List<ClientRequest>> unsent;
+        private final Map<Node, ConcurrentLinkedQueue<ClientRequest>> unsent;
 
         public UnsentRequests() {
             unsent = new HashMap<>();
         }
 
-        public synchronized void put(Node node, List<ClientRequest> requests) {
-            List<ClientRequest> nodeUnsent = unsent.get(node);
-            if (nodeUnsent == null) {
-                nodeUnsent = new ArrayList<>();
-                unsent.put(node, nodeUnsent);
-            }
-            nodeUnsent.addAll(requests);
-        }
-
         public synchronized void put(Node node, ClientRequest request) {
-            List<ClientRequest> nodeUnsent = unsent.get(node);
+            ConcurrentLinkedQueue<ClientRequest> nodeUnsent = unsent.get(node);
             if (nodeUnsent == null) {
-                nodeUnsent = new ArrayList<>();
+                nodeUnsent = new ConcurrentLinkedQueue<>();
                 unsent.put(node, nodeUnsent);
             }
             nodeUnsent.add(request);
         }
 
-        public synchronized int getRequestCount(Node node) {
-            List<ClientRequest> requests = unsent.get(node);
+        public synchronized int requestCount(Node node) {
+            ConcurrentLinkedQueue<ClientRequest> requests = unsent.get(node);
             return requests == null ? 0 : requests.size();
         }
 
-        public synchronized int getRequestCount() {
+        public synchronized int requestCount() {
             int total = 0;
-            for (List<ClientRequest> requests : unsent.values())
+            for (ConcurrentLinkedQueue<ClientRequest> requests : unsent.values())
                 total += requests.size();
             return total;
         }
 
         public synchronized boolean hasRequest(Node node) {
-            List<ClientRequest> requests = unsent.get(node);
+            ConcurrentLinkedQueue<ClientRequest> requests = unsent.get(node);
             return requests != null && !requests.isEmpty();
         }
 
         public synchronized boolean hasRequest() {
-            for (List<ClientRequest> requests : unsent.values())
+            for (ConcurrentLinkedQueue<ClientRequest> requests : unsent.values())
                 if (!requests.isEmpty())
                     return true;
             return false;
         }
 
-        public synchronized List<ClientRequest> removeExpiredRequests(long now, long unsentExpiryMs) {
-            List<ClientRequest> expiredRequests = new ArrayList<>();
-            Iterator<Map.Entry<Node, List<ClientRequest>>> iterator = unsent.entrySet().iterator();
+        public synchronized ConcurrentLinkedQueue<ClientRequest> removeExpiredRequests(long now, long unsentExpiryMs) {
```

Comment:
Sorry, my bad. LinkedIn organized a trip to Tahao today and I want to address Ismael's comments before the bus leaves at 5:30 pm. I was in a hurry to address all the comments and simply replaced many lists to `ConcurrentLinkedQueue`.

It is fixed now.